Improvement/cldsrv 724 sur utapi tests

benzekrimaha opened the pull request:
Issue: CLDSRV-724

bert-e commented:
# Hello benzekrimaha,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/cloudserver/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-724 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `9.1.1`


Please check the `Fix Version/s` of CLDSRV-724, or the target
branch of this pull request.

benzekrimaha committed:
utilities adaptation to comply with sdk migration

As test files depend on these utilities,
they needed to be adapted to match the sdk v3
requirements.

Issue: CLDSRV-724

benzekrimaha committed:
SUR tests sdk migration

Issue: CLDSRV-724

benzekrimaha committed:
UTAPI tests sdk migration

Issue: CLDSRV-724

benzekrimaha committed:
dependencies install

Issue: CLDSRV-724

williamlardier commented on tests/sur/quota.js:
```diff
@@ -339,8 +395,52 @@ function multiObjectDelete(bucket, keys, size, callback) {
+            });
+            
+            // Add middleware to strip ALL checksum headers
+            s3Client.middlewareStack.add(
```
Can we maybe add a comment to explain why we need this? And maybe a ticket number if that could be removed, so we just support the standard headers

williamlardier commented on tests/sur/quota.js:
```diff
@@ -196,27 +235,33 @@ function objectMPU(bucket, key, parts, partSize, callback) {
+            }
+            return callback(null, data);
+        })
+        .catch(err => {
+            callback(err);
+        });
```
```suggestion
        .catch(err => callback(err));
```

williamlardier commented on tests/sur/quota.js:
```diff
@@ -196,27 +235,33 @@ function objectMPU(bucket, key, parts, partSize, callback) {
-        if (!err && !s3Config.isQuotaInflightEnabled()) {
-            mockScuba.incrementBytesForBucket(bucket, parts * partSize);
+        if (!s3Config.isQuotaInflightEnabled()) {
+            mockScuba.incrementBytesForBucket(bucket, -(parts * partSize));
```
i'm not sure to get why we increment the bytes before the call then we decrement them, is there any issue with the test?

williamlardier commented on tests/sur/quota.js:
```diff
@@ -57,109 +77,136 @@ function putObjectLockConfiguration(bucket, cb) {
+            }
+            return cb(null, data);
+        })
+        .catch(err => {
+            cb(err);
+        });
```
```suggestion
        .catch(err => cb(err));
```

DarkIsDude replied:
`catch(cb)

williamlardier reviewed (commented):
I just have one comment "blocking", it's the behavior change on one of the quota tests, i'm not sure this is needed

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-724 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `9.1.4`


Please check the `Fix Version/s` of CLDSRV-724, or the target
branch of this pull request.

DarkIsDude commented on tests/functional/aws-node-sdk/lib/utility/bucket-util.js:
```diff
@@ -1,139 +1,146 @@
-        this.s3.config.setPromisesDependency(Promise);
-        this.s3.config.update({
-            maxRetries: 0,
-        });
+        if (unauthenticated) {
+            this.s3 = new S3Client({
+                ...s3Config,
+                credentials: { accessKeyId: '', secretAccessKey: '' },
+                forcePathStyle: true,
+                signer: { sign: async request => request },
+             });
+        }
+       else {
+            this.s3 = new S3Client({
+                ...s3Config,
+                maxAttempts: 0,
+             });
+       }
```
```suggestion
        if (unauthenticated) {
            s3Config = {
              ...s3Config,
              credentials: { accessKeyId: '', secretAccessKey: '' },
              forcePathStyle: true,
              signer: { sign: async request => request },
            };
        }

            this.s3 = new S3Client({
                ...s3Config,
                maxAttempts: 0,
             });
```

Pretty sure that `maxAttempts: 0,` is working also for unauthenticated client ?

DarkIsDude commented on tests/functional/aws-node-sdk/lib/utility/bucket-util.js:
```diff
@@ -1,139 +1,146 @@
+        return this.s3.send(new DeleteBucketCommand({ Bucket: bucketName }))
+            .catch(err => {
+                throw err;
+            });
```
```suggestion
        return await this.s3.send(new DeleteBucketCommand({ Bucket: bucketName }));
```

One good practice when working with async/await is to always await with return, this will help to preserve the full stack trace (without the await, the stack trace will start at the caller, if you put the await here, the stack trace will display `deleteOne`

benzekrimaha replied:
this is indeed accurate when the function is async , which is not the case here , in case of an error the stack trace will still show the error which is what we want / also these functions are used in tests only => we'll be able to see the api logs in case of debug. Still the catch is unnecessary so removing it

DarkIsDude commented on tests/functional/aws-node-sdk/lib/utility/bucket-util.js:
```diff
@@ -1,139 +1,146 @@
+                    )
+                    .concat((data.DeleteMarkers || [])
+                        .map(object =>
+                        this.s3.send(new DeleteObjectCommand({
```
This indent is weird ?

DarkIsDude commented on tests/functional/aws-node-sdk/lib/utility/bucket-util.js:
```diff
@@ -1,139 +1,146 @@
     emptyMany(bucketNames) {
-        const promises = bucketNames.map(
-            bucketName => this.empty(bucketName)
+            const promises = bucketNames.map(
```
Indent issue ?

DarkIsDude commented on tests/functional/aws-node-sdk/lib/utility/bucket-util.js:
```diff
@@ -1,139 +1,146 @@
         return Promise.all(promises);
     }
-
     emptyIfExists(bucketName) {
```
I think our guidelines ask us to add a new line between function definition ?

benzekrimaha replied:
good catch missed it here

DarkIsDude commented on tests/functional/aws-node-sdk/test/support/config.js:
```diff
@@ -58,11 +85,11 @@ function _getAwsConfig(profile, config) {
-
-    return fn.apply(this, [profile, config]);
+function getConfig(profile, config) {
+    if (process.env.AWS_ON_AIR) {
```
Are you sure about that ? This one is essential `process.env.AWS_ON_AIR === 'true' as `AWS_ON_AIR = false` will consider it as `AWS_ON_AIR` in your case. An env var is always a string

benzekrimaha replied:
Actually it's either set or not at all this is the check we have across the whole code base so I'd rather keep it

DarkIsDude commented on tests/utapi/awsNodeSdk.js:
```diff
@@ -1,159 +1,194 @@
+
+            const headers = args.request.headers;
+
+            // Remove any checksum headers
```
```suggestion
```

DarkIsDude commented on tests/utapi/awsNodeSdk.js:
```diff
@@ -1,159 +1,194 @@
+                }
+            });
+
+            // Add MD5
+            if (args.request.body) {
+                const bodyContent = Buffer.from(args.request.body);
+                // Create a new hash instance for each request
+                headers['Content-MD5'] = createHash('md5').update(bodyContent).digest('base64');
+            }
```
```suggestion
            if (args.request.body) {
                const bodyContent = Buffer.from(args.request.body);
                headers['Content-MD5'] = createHash('md5').update(bodyContent).digest('base64');
            }
```

benzekrimaha replied:
with the latest fixes made by leif on the api level we don't need this anymore so it has been removed as we only validate the md5 when provided now

DarkIsDude commented on tests/utapi/awsNodeSdk.js:
```diff
@@ -1,159 +1,194 @@
+ * Creates an S3 client that uses MD5 checksums for DeleteObjects operations
+ * Based on https://github.com/aws/aws-sdk-js-v3/blob/main/supplemental-docs/MD5_FALLBACK.md
+ */
+function createS3ClientWithMD5(configuration = {}) {
```
```suggestion
function createS3ClientWithMD5Checksum(configuration = {}) {
```

nits

DarkIsDude commented on tests/utapi/awsNodeSdk.js:
```diff
@@ -1,159 +1,194 @@
-        assert.ifError(err);
-        return cb(err, data);
-    });
+    return s3Client.send(new CreateBucketCommand({ Bucket: bucket }))
```
Do we need to return ? Seems that we use only the cb here ?

DarkIsDude commented on tests/utapi/awsNodeSdk.js:
```diff
@@ -1,159 +1,194 @@
-        return cb(err);
-    });
+    return s3Client.send(new DeleteObjectCommand({ Bucket: bucket, Key: key }))
+        .then(() => cb())
```
```suggestion
        .then(cb)
```

benzekrimaha replied:
.then(cb) would be problematic because it doesn't follow the error-first callback convention that Node.js expects.

DarkIsDude commented on tests/utapi/awsNodeSdk.js:
```diff
@@ -1,159 +1,194 @@
-        return cb(err);
-    });
+    return s3Client.send(new DeleteBucketCommand({ Bucket: bucket }))
+        .then(() => cb())
```
```suggestion
        .then(cb)
```

DarkIsDude commented on tests/utapi/awsNodeSdk.js:
```diff
@@ -1,159 +1,194 @@
-        return cb(err);
-    });
+    return s3Client.send(new DeleteObjectsCommand(params))
+        .then(() => cb())
```
```suggestion
        .then(cb)
```

DarkIsDude commented on tests/utapi/awsNodeSdk.js:
```diff
@@ -1,159 +1,194 @@
-    });
+    const params = { Bucket: bucket, CopySource: `${bucket}/${key}`, Key: `${key}-copy` };
+    return s3Client.send(new CopyObjectCommand(params))
+        .then(() => cb())
```
```suggestion
        .then(cb)
```

Same for other iterations

DarkIsDude commented on tests/utapi/awsNodeSdk.js:
```diff
@@ -1,159 +1,194 @@
-            err => cb(err, data)),
-        (data, cb) => deleteVersionList(data.Versions, bucket,
-            err => cb(err, data)),
+        cb => s3Client.send(new ListObjectVersionsCommand(params))
```
```suggestion
        async() => s3Client.send(new ListObjectVersionsCommand(params)),
```

Async can understand the async/await pattern

DarkIsDude commented on tests/utapi/awsNodeSdk.js:
```diff
@@ -162,33 +197,28 @@ function objectMPU(bucket, key, parts, partSize, callback) {
-        assert.ifError(err);
-        return cb(err, data);
-    });
+    return s3Client.send(new GetObjectCommand({ Bucket: bucket, Key: key }))
```
Same it's maybe not needed to return ?

DarkIsDude commented on tests/sur/quota.js:
```diff
@@ -27,26 +43,30 @@ function createBucket(bucket, locked, cb) {
+    return s3Client.send(new CreateBucketCommand(config)) 
+        .then(data => cb(null, data))
+        .catch(err => {
+            assert.ifError(err);
```
I was curious about an assert in a catch and I think this will not work:

```
const assert = require('assert');

describe.only('Array', () => {
  it.only('should return -1 when the value is not present', done => {
    function test(cb) {
      new Promise((_resolve, reject) => {
        reject('C est pas vrai');
      })
        .then(cb)
        .catch(err => {
          assert.ifError(err);
          cb();
        });
    }

    test(done);
  });
});
```

The callback is never called ?

DarkIsDude commented on tests/sur/quota.js:
```diff
@@ -57,109 +77,136 @@ function putObjectLockConfiguration(bucket, cb) {
+            }
+            return cb(null, data);
+        })
+        .catch(err => {
+            cb(err);
+        });
```
```suggestion
        .catch(cb(err));
```

DarkIsDude commented on tests/sur/quota.js:
```diff
@@ -57,109 +77,136 @@ function putObjectLockConfiguration(bucket, cb) {
-    request.on('build', () => {
-        request.httpRequest.headers['x-scal-s3-version-id'] = vID;
-    });
+    // Add custom header using middleware
```
```suggestion
```

DarkIsDude commented on tests/sur/quota.js:
```diff
@@ -57,109 +77,136 @@ function putObjectLockConfiguration(bucket, cb) {
+            }
+            cb(null, data);
+        })
+        .catch(err => cb(err));
```
```suggestion
        .catch(cb);
```

DarkIsDude commented on tests/sur/quota.js:
```diff
@@ -296,35 +339,48 @@ function uploadPartCopy(bucket, key, partNumber, partSize, sleepDuration, keyToC
+            callback(null, data);
+        })
+        .catch(err => {
+
+            if (!s3Config.isQuotaInflightEnabled()) {
+                mockScuba.incrementBytesForBucket(bucket, size);
+            }
+            return callback(err);
+        });
```
```suggestion
        .catch(err => {
            if (!s3Config.isQuotaInflightEnabled()) {
                mockScuba.incrementBytesForBucket(bucket, size);
            }
            
            return callback(err);
        });
```

nits

DarkIsDude commented on tests/sur/quota.js:
```diff
@@ -361,10 +461,12 @@ function multiObjectDelete(bucket, keys, size, callback) {
-                }),
+                next => {
+                    putObject(bucket, key, size, err => {
+                        assert.strictEqual(err.Code, 'QuotaExceeded');
```
Same here, I'm not sure the callback will be called ?

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-724 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `9.1.5`


Please check the `Fix Version/s` of CLDSRV-724, or the target
branch of this pull request.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-724 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `9.1.5`


Please check the `Fix Version/s` of CLDSRV-724, or the target
branch of this pull request.

SylvainSenechal commented on tests/functional/aws-node-sdk/lib/utility/bucket-util.js:
```diff
@@ -1,81 +1,95 @@
+        if (unauthenticated) {
+            this.s3 = new S3Client({
+                ...s3Config,
+                maxAttempts: 0,
```
Should be 1, although I guess probably 0 is interpreted as 1 too

benzekrimaha replied:
0 is indeed interpreted as 1 so keeping it that way as it's working across all tests

SylvainSenechal commented on tests/sur/quota.js:
```diff
@@ -248,30 +269,28 @@ function uploadPartCopy(bucket, key, partNumber, partSize, sleepDuration, keyToC
-            });
+            return s3Client.send(new UploadPartCommand(uploadPartParams))
+                .then(data => {
+                    ETags[partNumber] = data.ETag;
```
Just wanna make sure :
In the function UploadPartCopyCommand, you used
`data.CopyPartResult.ETag`
Are you sure that here, its not something like 
`data.UploadPartResult.ETag` ?

benzekrimaha replied:
Based on the documentation and tests passing it's directly retrieved with .Etag no object encapsulation here => https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/Package/-aws-sdk-client-s3/Interface/UploadPartCommandOutput/

SylvainSenechal reviewed (commented):
Looks good to me, I'll let you address the last comments and Edouard approve it

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers






The following reviewers are expecting changes from the author, or must review again:

* @darkisdude

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-724 contains:


* `9.3.0`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `9.1.7`

* `9.2.5`


Please check the `Fix Version/s` of CLDSRV-724, or the target
branch of this pull request.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-724 contains:


* `9.3.0`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `9.1.7`

* `9.2.6`


Please check the `Fix Version/s` of CLDSRV-724, or the target
branch of this pull request.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-724 contains:


* `9.3.0`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `9.2.6`


Please check the `Fix Version/s` of CLDSRV-724, or the target
branch of this pull request.
